src/utilities/common/canonicalStringify.ts

```diff
@@ -0,0 +1,81 @@
+ * significant role, as `canonicalStringify` will be called for only a limited
+ * number of object shapes, and the cache will not grow beyond a certain point.
+ * But in some edge cases, this could be a problem, so we provide
+ * canonicalStringify.reset() as a way of clearing the cache.
+ * */
+export const canonicalStringify = Object.assign(
+  function canonicalStringify(value: any): string {
+    return JSON.stringify(value, stableObjectReplacer);
+  },
+  {
+    reset() {
+      // Clearing the sortingMap will reclaim all cached memory, without
+      // affecting the logical results of canonicalStringify, but potentially
+      // sacrificing performance until the cache is refilled.
+      sortingMap.clear();
+    },
+  }
+);
+
+// Values are JSON-serialized arrays of object keys (in any order), and values
+// are sorted arrays of the same keys.
+const sortingMap = new Map<string, readonly string[]>();
+
+// The JSON.stringify function takes an optional second argument called a
+// replacer function. This function is called for each key-value pair in the
+// object being stringified, and its return value is used instead of the
+// original value. If the replacer function returns a new value, that value is
+// stringified as JSON instead of the original value of the property.
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#the_replacer_parameter
+function stableObjectReplacer(key: string, value: any) {
+  if (value && typeof value === "object") {
+    const proto = Object.getPrototypeOf(value);
+    // We don't want to mess with objects that are not "plain" objects, which
+    // means their prototype is either Object.prototype or null. This check also
+    // prevents needlessly rearranging the indices of arrays.
+    if (proto === Object.prototype || proto === null) {
+      const keys = Object.keys(value);
+      if (isArraySorted(keys)) return value;
+      const unsortedKey = JSON.stringify(keys);
+      let sortedKeys = sortingMap.get(unsortedKey);
+      if (!sortedKeys) {
+        keys.sort();
+        const sortedKey = JSON.stringify(keys);
+        // Checking for sortedKey in the sortingMap allows us to share the same
+        // sorted array reference for all permutations of the same set of keys.
+        sortedKeys = sortingMap.get(sortedKey) || keys;
+        sortingMap.set(unsortedKey, sortedKeys);
+        sortingMap.set(sortedKey, sortedKeys);
+      }
+      const sortedObject = Object.create(proto);
+      // Reassigning the keys in sorted order will cause JSON.stringify to
+      // serialize them in sorted order.
+      sortedKeys.forEach((key) => {
+        sortedObject[key] = value[key];
+      });
+      return sortedObject;
+    }
+  }
+  return value;
+}
+
+function isArraySorted(keys: readonly string[]): boolean {
+  for (let k = 1, len = keys.length; k < len; ++k) {
+    if (keys[k - 1] > keys[k]) {
+      return false;
+    }
+  }
+  return true;
```

Comment:
```suggestion
return !keys.some((value, index) => index > 0 && keys[index - 1] > value);
```

This one could save us 18 bytes here - I think it should not be significantly faster/slower in any way.

Reply:
I almost had assumed something like that :)
I'll approve the PR then 👍

Reply:
Found a way to avoid any allocations: d0aa0301ebfaf5b6894386382a4a2c2c7a350fb6

Reply:
Great!